### PR TITLE
feat: Add option to disable previous URL storage

### LINKED ISSUE #81: bug: Saving the previous URL
Otherwise, the HTMX request may change the previous URL. While we are updating on the same page (tab), there is no problem. But if `previous_url() `returns another page, the result is unexpected.
File **[vendor/codeigniter4/framework/system/CodeIgniter.php](https://github.com/codeigniter4/CodeIgniter4/blob/develop/system/CodeIgniter.php)** in `storePreviousURL()`:
```php
// Ignore AJAX requests
// Possible solution
// if ((method_exists($this->request, 'isHTMX') && $this->request->isHTMX()) || (method_exists($this->request, 'isAJAX') && $this->request->isAJAX())) {
if (method_exists($this->request, 'isAJAX') && $this->request->isAJAX()) {
    return;
}
...
```
or identify HTMX = AJAX. But it can introduce new dependencies.

### PR DESCRIPTION
**Pull request by neznaika0**
Closes #81 
Copied `isWeb()` - it private method
The test may be inaccurate, correct it

### CODE REVIEW

**Comment by michalsn** [comment]
I have second thoughts on this topic. When working with htmx, we may often want the page loaded by the htmx request to be stored as the previous URL.

My point is that in some cases this can be helpful, but for others, it will be a problem... so IMO we should make this option configurable.

Maybe `$storePreviousURL` option in the config file, with the default value set to `true` to preserve BC?

**Comment by neznaika0** [comment]
An example of specific cases? Yes, the setup is not a problem
What about the AJAX option? It does not save by default. If the HTMX is configured with AJAX headers, the option has no effect.

**Comment by michalsn** [comment]
> An example of specific cases?

An example would be using htmx in the navigation to simply navigate to another page. In that case, we won't be storing the previous URL in the session at all.

When AJAX headers are used, it is the user's choice. We will not override the default behavior of the framework.

**Comment by neznaika0** [comment]
No 😄 . An example where the previous URL is needed.
In the **userform > formhandler > userform** chain, sometimes the previous URL may be **formhandler**. 
I want to avoid this. 

Plus, there are questions about the framework when working with browser tabs... there, too, the URL is overwritten.

I'll update the PR tomorrow. I would be grateful if we discuss it in more detail.

**Comment by michalsn** [comment]
It is not important where it will be used, only whether it will return correct data. If you disable updating the previous URL for htmx requests, then it may have an incorrect value.

> Plus, there are questions about the framework when working with browser tabs... there, too, the URL is overwritten.

Yes, the same way as for traditional requests - we won't change that.

File: docs/configuration.md [hunk 1]
```diff
@@ -36,3 +36,10 @@ Since these decorators are used automatically in the `development` mode (or to b
 We can add the defined string as an `id` or `class` to the html tag.
 
 In the `production` environment these decorators are ignored by design. So this is useful only for the `development` mode.
+
+### $storePreviousURL
+
+Saving the previous URL between requests.
+Sometimes this breaks the logic - the previous URL is part of the HTMX instead of the request page.
+
+In the **/home > /htmx-request > /home** query chain, you can sometimes get an unexpected result if the previous URL is saved between requests. Instead of **/home**, you will get **/htmx-request**. The value `false` will save only page from which request was received
```
Comment (michalsn) [inline docs/configuration.md]:
```suggestion
Specifies whether the HTMX request URL should be stored in the session, for use with the `previous_url()` helper function.
For more information, see the [user guide](https://codeigniter.com/user_guide/helpers/url_helper.html#previous_url).

Basically, if you use HTMX extensively, including for navigating your site, you will probably want to leave it as `true`,
and in cases where storing the request is not desirable, even if it uses HTMX, you can use custom header, to indicate the
AJAX call or [ajax-header](https://github.com/bigskysoftware/htmx-extensions/blob/main/src/ajax-header/README.md) extension,
which will add the necessary headers automatically. URLs from AJAX requests are always excluded from session storage.
```

File: src/CodeIgniter.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+use CodeIgniter\HTTP\DownloadResponse;
+use CodeIgniter\HTTP\IncomingRequest;
+use CodeIgniter\HTTP\RedirectResponse;
+use CodeIgniter\HTTP\URI;
+use Michalsn\CodeIgniterHtmx\Config\Htmx;
+use Michalsn\CodeIgniterHtmx\HTTP\IncomingRequest as HTMXIncomingRequest;
+
+/**
+ * @property CLIRequest|HTMXIncomingRequest|IncomingRequest|null $request
+ */
```
Comment (michalsn) [inline src/CodeIgniter.php]:
```suggestion
use CodeIgniter\CodeIgniter as BaseCodeIgniter;
use CodeIgniter\HTTP\CLIRequest;
use CodeIgniter\HTTP\DownloadResponse;
use CodeIgniter\HTTP\RedirectResponse;
use CodeIgniter\HTTP\URI;
use Michalsn\CodeIgniterHtmx\HTTP\IncomingRequest;

/**
 * @property CLIRequest|IncomingRequest|null $request
 */
```

File: src/CodeIgniter.php [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+        $storePreviousURL = config(Htmx::class)->storePreviousURL;
+
+        // Ignore AJAX and HTMX requests
+        if (
+            ($storePreviousURL === false && (method_exists($this->request, 'isHTMX') && $this->request->isHTMX()))
+            || (method_exists($this->request, 'isAJAX') && $this->request->isAJAX())
+        ) {
+            return;
+        }
```
Comment (michalsn) [inline src/CodeIgniter.php]:
```suggestion
        // Ignore AJAX requests
        if (method_exists($this->request, 'isAJAX') && $this->request->isAJAX()) {
            return;
        }

        // Ignore HTMX requests
        if (config('Htmx')->storePreviousURL === false
            && method_exists($this->request, 'isHTMX') && $this->request->isHTMX()) {
            return;
        }
```

File: src/Config/Htmx.php [hunk 4]
```diff
@@ -22,4 +22,10 @@ class Htmx extends BaseConfig
      * when they are enabled.
      */
     public string $skipViewDecoratorsString = 'htmxSkipViewDecorators';
+
+    /**
+     * Saving the previous URL between requests
+     * Sometimes this breaks the logic - the previous URL is part of the HTMX instead of the request page
+     */
```
Comment (michalsn) [inline src/Config/Htmx.php]:
```suggestion
    /**
     * Enable / disable storing previous URLs for HTMX requests.
     */
```

File: tests/CodeIgniterTest.php [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    protected function tearDown(): void
+    {
+        parent::tearDown();
+
+        $this->resetServices();
+    }
+
+    public function testIsHTMXNotSavePreviousURL(): void
```
Comment (michalsn) [inline tests/CodeIgniterTest.php]: Please fix the test name and split it, so we test one scenario at a time.

**Review by michalsn — CHANGES_REQUESTED** [review]

**Comment by neznaika0** [comment]
You have almost completely corrected the PR. Fix it faster for you

**Review by michalsn — APPROVED** [review]
You did 90% of the work here. Thank you.

**Comment by neznaika0** [comment]
@michalsn I caught BC :smile: 
Let's say there is a "Show deleted" checkbox in the list of users. After clicking, HTMX changes the URI in the browser, but not in the history. If we click another button/link and try to return to the current URI (as `previous_url()`), it will turn out to be the old one (before clicking the "Show deleted" button)
Now I fully understand your pagination example.

So I haven't completely solved my problem.

**Comment by michalsn** [comment]
What is your problem? Sometimes you want to store the `previousURL` even if it's disabled?
Or are you referring to the browser history?

**Comment by neznaika0** [comment]
The example with pagination is simpler: 
1. Go to **/users**
2. Click to pages Next/Back from HTMX and stop at **/users?page=10**
3. Click on the button with HTMX **/users/ban/33** and return to the same page (via `previous_url()`) Ops.. We should go back to **/users?page=10** but the story is disabled - we get to **/users**

Now I have forgotten the exact example of the initial problem. But apparently this is not how it was planned.

EDIT: A good solution is to use `HX-Current-Url` as it sends the URL of the start page. It is similar to the Referer but looks more reliable

**Comment by michalsn** [comment]
Why not use `ajax-header` extension for places where you ~want~ don't want to store the URL?

**Comment by neznaika0** [comment]
Let's leave this question for now. When I find the exact example, I'll come back.

**Pull request merged**